### DISCUSSION TITLE
Improving magnetic field calculations for non-symmetric setups

### DISCUSSION THREAD

**Discussion by renereimann** [KATRIN-Experiment/Kassiopeia#99] [Ideas]
Hej all,
within Project 8 we are experience performance issues for setups that are not axial symmetric.
The reason is that no magnetic field solver except direct numerical Biot-Savart law integration is implemented in Kassiopeia for setups that are not axis symmetric.

We come up with three ideas on how non-axial symmetric setups can be implemented.

1. Implementation of the Fast Fourier Transform on Multipoles (FFTM) that has been already implemented for E-fields
2. Upgrade the ZonalHarmonic solver to a SphericalHarmonic solver
3. Allow to import unstructured grids.

Details can be found in the presentation that is attached and was presented at the Kassiopeia User-Meeting on Feb 26, 2024.

[P8_Kassiopeia_UserMeeting_2024_02_26.pdf](https://github.com/KATRIN-Experiment/Kassiopeia/files/14421877/P8_Kassiopeia_UserMeeting_2024_02_26.pdf)

**Comment by renereimann** [comment]
Hej all,

I just found by accident that VTK supports the usage of unstructured grids, see e.g. here https://examples.vtk.org/site/Python/IO/ReadUnstructuredGrid/
That would make the IO part of option 3 relatively simple since we already have the VTK support build in. The part that would be needed is the interpolation of the field given the grid.

I also wondered, if the "surface charges" on the FFTM can be exported / imported based on an unstructured grid as well. While above for option 3 the unstructured grid would be a volume grid (3D), in the surface charges of the FFTM it would be a surface grid (2D).

Happy to take any feedback on those ideas.

**Reply by richeldichel** [reply]
That sounds great. From a naive point of view, this looks like the easiest solution now. Looking forward to seeing more progress on this!
